Improve isolation tests for metadata syncing

hanefi opened the pull request:
This PR aims to cover some concurrent operations with metadata syncing.

The list of operations are tracked in https://github.com/citusdata/citus/issues/1199

I added tests that capture the following classes of queries
- [x] concurrent operations with `start_metadata_syncing` that are blocked
	- [x] `ALTER TABLE`
	- [x] `DROP TABLE`
	- [x] `create_distributed_table()`
	- [x] `create_reference_table()`
	- [x] `ATTACH PARTITION`
	- [x] `DETACH PARTITION`
	- [x] `CREATE TABLE .. PARTITION OF`
	- [x] Adding foreign keys from dist to ref tables
	- [x] Dropping foreign keys from dist to ref tables
- [x] operations that are not get blocked.
	- [x] creating distributed types
	- [x] Creating distributed functions
- [x] Operations that are cancelled by distributed deadlock detection
	- [x] `DROP TYPE` commands for distributed types.

TODO:
- [ ] Investigate the root cause of the deadlocks above
- [ ] Try to reproduce #4366 and document findings.

onderkalaci commented:
Hi @hanefi,

Thanks, the tests you added looks good. On top of the missing items you have as TODO, we could consider the following as well:

Few more ideas could be:
- create distributed function
- create foreign key to a distributed/reference table
- Can you check if this is fixed: https://github.com/citusdata/citus/issues/4366 and if so, add test

onderkalaci commented on src/test/regress/spec/isolation_metadata_sync_vs_all.spec:
```diff
@@ -0,0 +1,117 @@
+	SET citus.enable_metadata_sync_by_default TO 'on';
+	SELECT 1 FROM master_add_node('localhost', 57638);
+
+	SELECT public.wait_until_metadata_sync(5000);
```
metadata sync in synchronous so probably not needed

onderkalaci commented on src/test/regress/expected/isolation_metadata_sync_vs_all.out:
```diff
@@ -0,0 +1,595 @@
+Parsed test spec with 3 sessions
+
+starting permutation: s1-begin s2-begin s1-start-metadata-sync s2-alter-table s1-rollback s2-rollback s3-compare-snapshot
+create_distributed_function
```
I think not a critical thing, but seeing `create_distributed_function` on each permutation felt unexpected at first. I guess, it shows up because it is the last step of the setup.  Can we somehow  avoid this?

hanefi replied:
I went over the statements in the setup phase, and the outputs are more clear now.

onderkalaci commented on src/test/regress/expected/isolation_metadata_sync_vs_all.out:
```diff
@@ -0,0 +1,595 @@
+ SELECT create_distributed_table('new_dist_table', 'id');
+ <waiting ...>
+step s1-rollback: 
+    ROLLBACK;
```
Why do we ROLLBACK in the tests? Wouldn't it be better to `COMMIT` and `s3-compare-snapshot`? Such that we ensure the current operation is done successfully? '

Valid for all rollback steps in the test -- unless there are any specific ROLLBACKs

hanefi replied:
I commit everything now.

onderkalaci commented on src/test/regress/spec/isolation_metadata_sync_vs_all.spec:
```diff
@@ -0,0 +1,197 @@
+
+step "s3-compare-snapshot"
+{
+	SELECT bool_and(result::text[] @> activate_node_snapshot() AND result::text[] <@ activate_node_snapshot()) AS same_snapshot_across_cluster
```
Not sure if my below suggestion is simpler than this, but at least I find it easier to read and seems does a full line by line comparison across all nodes:
```SQL

SELECT count(*) = 0
FROM 
(
    (
    	SELECT unnest(activate_node_snapshot())
     		EXCEPT 
     	SELECT unnest(string_to_array(RESULT, 'CANNOTBEUSEDSTRING')) AS unnested_result
     	FROM run_command_on_workers($$SELECT array_to_string(activate_node_snapshot(), 'CANNOTBEUSEDSTRING')$$)
     	GROUP BY unnested_result
     ) 
  UNION
    (
    	SELECT unnest(string_to_array(RESULT, 'CANNOTBEUSEDSTRING')) AS unnested_result
     	FROM run_command_on_workers($$SELECT array_to_string(activate_node_snapshot(), 'CANNOTBEUSEDSTRING')$$)
     	GROUP BY unnested_result
  			EXCEPT
    	SELECT unnest(activate_node_snapshot())
    )
) AS foo;
```

One difference I found is that the coordinator generates more lines for things like:
```SQL
 SET citus.enable_ddl_propagation TO 'on'
 SET citus.enable_ddl_propagation TO 'off'
 SET ROLE onderkalaci
```

I think that is a separate discussion, the above query still does a full control for

hanefi replied:
I simplified your suggested queries. Let me know if you think they are ok.

Basically, I removed conversions from text[] array into string, and vice-versa. We can already use text[] values returned by the `run_command_on_workers()` UDF.

onderkalaci commented on src/test/regress/spec/isolation_metadata_sync_vs_all.spec:
```diff
@@ -0,0 +1,197 @@
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-type" "s1-commit" "s2-commit" "s3-compare-snapshot" "s2-drop-type"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-dist-func" "s1-commit" "s2-commit" "s3-compare-snapshot" "s2-drop-dist-func"
+
+// the following operation creates a distributed deadlock and gets cancelled
```
I'll check this

hanefi replied:
I added one more deadlock case when dropping a distributed function concurrent with a `start_metadata_sync` operation.

hanefi commented:
I now have a broken isolation test `isolation_create_table_vs_add_remove_node` that I am trying to fix.

Permutation: `s1-add-node-2 s2-begin s2-create-append-table s1-remove-node-2 s2-commit s2-select`

```diff
 step s1-remove-node-2: <... completed>
-ERROR:  cannot remove or disable the node localhost:xxxxx because because it contains the only shard placement for shard xxxxx
+master_remove_node
+---------------------------------------------------------------------
+
+(1 row)
+
```

onderkalaci commented on src/test/regress/spec/isolation_metadata_sync_vs_all.spec:
```diff
@@ -0,0 +1,205 @@
+	ALTER SEQUENCE pg_catalog.pg_dist_node_nodeid_seq RESTART 123000;
+	ALTER SEQUENCE pg_catalog.pg_dist_shardid_seq RESTART 123000;
+
+	SELECT 1 FROM master_add_node('localhost', 57637);
```
why do we add / remove nodes in the setup / tear down? Are they strictly necessary?

hanefi replied:
No they are not. I used to have one node with metadata, and one without. This is no longer the case now.

My original plan was to test `citus_add_node`, `start_metadata_sync_to_node`, `trigger_metadata_sync` and I no longer need to add and drop the nodes at every permutation.

onderkalaci commented on src/test/regress/spec/isolation_metadata_sync_vs_all.spec:
```diff
@@ -0,0 +1,205 @@
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-type" "s1-commit" "s2-commit" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-dist-func" "s1-commit" "s2-commit" "s3-compare-snapshot"
+
+// the following operations create a distributed deadlock and gets cancelled
```
my suggestion is to remove these two tests from this PR and merge this. And then dive into why DROP objects fails in general

hanefi replied:
For future reference, here are the permutations that caused deadlocks that I removed from the file:

```
// the following operations create a distributed deadlock and gets cancelled
permutation "s2-create-type" "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-drop-type" "s1-commit" "s2-commit" "s3-compare-snapshot"
permutation "s2-create-dist-func" "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-drop-dist-func" "s1-commit" "s2-commit" "s3-compare-snapshot"
```

onderkalaci commented on src/test/regress/spec/isolation_metadata_sync_vs_all.spec:
```diff
@@ -0,0 +1,205 @@
+
+step "s2-attach-partition"
+{
+	ALTER TABLE dist_partitioned_table ATTACH PARTITION dist_partitioned_table_p1 FOR VALUES FROM (1) TO (9);
```
one final request: Can you also please add `CREATE PARTITION OF` test? That goes from a slightly different code-path, so would be good to have that

hanefi replied:
We already cover this. See `s2-create-partition-of` step.

onderkalaci commented on src/test/regress/spec/isolation_metadata_sync_vs_all.spec:
```diff
@@ -0,0 +1,205 @@
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-drop-table" "s1-commit" "s2-commit" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-dist-table" "s1-commit" "s2-commit" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-create-ref-table" "s1-commit" "s2-commit" "s3-compare-snapshot"
+permutation "s1-begin" "s2-begin" "s1-start-metadata-sync" "s2-attach-partition" "s1-commit" "s2-commit" "s3-compare-snapshot"
```
should we add "s1-start-metadata-sync" vs "s2-start-metadata-sync" as well? We might have such test in another file. If so, we can skip addinng that here

hanefi replied:
I created another section above for these tests.

onderkalaci reviewed (approved):
If you agree with my comments, we can merge and continue with DROP bug(s). Otherwise, let's discuss further

hanefi commented:
@onderkalaci I addressed your comments, and I did some more changes:

1. I moved the test to later in the schedule as I no longer need to add or remove nodes.
2. I added more permutations with comments so that it will be easier to understand how to fix the tests if they fail. For example, I check for metadata consistency before running any of the queries so if an earlier test failed and broke metadata, it is easier to understand that.

Please take another quick look before I merge this one.
